File: packages/react-native-autoplay/ios/hybrid/HybridNowPlayingTemplate.swift
//
//  HybridNowPlayingTemplate.swift
//  react-native-autoplay
//
//  Created by Claude Code
//

import NitroModules

class HybridNowPlayingTemplate: HybridNowPlayingTemplateSpec {
    func configureNowPlayingTemplate(config: NowPlayingTemplateConfig) throws {
        let template = NowPlayingTemplate(config: config)
        TemplateStore.addTemplate(
            template: template,
            templateId: config.id
        )
    }

    func updateNowPlayingButtons(
        templateId: String,
        buttons: [NitroNowPlayingButton]
    ) throws -> Promise<Void> {
        return Promise.async {
            guard
                let template = TemplateStore.getTemplate(templateId: templateId)
                    as? NowPlayingTemplate
            else {
                throw AutoPlayError.invalidTemplateType(
                    "\(templateId) is not a NowPlayingTemplate"
                )
            }

            await template.updateButtons(buttons: buttons)
        }
    }

    func setAlbumArtistButtonEnabled(
        templateId: String,
        enabled: Bool
    ) throws -> Promise<Void> {
        return Promise.async {
            guard
                let template = TemplateStore.getTemplate(templateId: templateId)
                    as? NowPlayingTemplate
            else {
                throw AutoPlayError.invalidTemplateType(
                    "\(templateId) is not a NowPlayingTemplate"
                )
            }

            await template.setAlbumArtistButtonEnabled(enabled: enabled)
        }
    }

    func setUpNextButtonEnabled(
        templateId: String,
        enabled: Bool,
        title: String?
    ) throws -> Promise<Void> {
        return Promise.async {
            guard
                let template = TemplateStore.getTemplate(templateId: templateId)
                    as? NowPlayingTemplate
            else {
                throw AutoPlayError.invalidTemplateType(
                    "\(templateId) is not a NowPlayingTemplate"
                )
            }

            await template.setUpNextButtonEnabled(enabled: enabled, title: title)
        }
    }
}

File: packages/react-native-autoplay/android/src/main/java/com/margelo/nitro/swe/iternio/reactnativeautoplay/HybridNowPlayingTemplate.kt
package com.margelo.nitro.swe.iternio.reactnativeautoplay

import com.margelo.nitro.core.Promise
import com.margelo.nitro.swe.iternio.reactnativeautoplay.template.AndroidAutoTemplate
import com.margelo.nitro.swe.iternio.reactnativeautoplay.template.NowPlayingTemplate

class HybridNowPlayingTemplate : HybridNowPlayingTemplateSpec() {

    override fun configureNowPlayingTemplate(config: NowPlayingTemplateConfig) {
        val context = AndroidAutoSession.getRootContext()
            ?: throw IllegalArgumentException("configureNowPlayingTemplate failed, carContext not found")

        val template = NowPlayingTemplate(context, config)
        AndroidAutoTemplate.setTemplate(config.id, template)
    }

    override fun updateNowPlayingButtons(
        templateId: String,
        buttons: Array<NitroNowPlayingButton>
    ): Promise<Unit> {
        return Promise.async {
            val template = AndroidAutoTemplate.getTemplate<NowPlayingTemplate>(templateId)
            template.updateButtons(buttons)
        }
    }

    override fun setAlbumArtistButtonEnabled(
        templateId: String,
        enabled: Boolean
    ): Promise<Unit> {
        return Promise.async {
            val template = AndroidAutoTemplate.getTemplate<NowPlayingTemplate>(templateId)
            template.setAlbumArtistButtonEnabled(enabled)
        }
    }

    override fun setUpNextButtonEnabled(
        templateId: String,
        enabled: Boolean,
        title: String?
    ): Promise<Unit> {
        return Promise.async {
            val template = AndroidAutoTemplate.getTemplate<NowPlayingTemplate>(templateId)
            template.setUpNextButtonEnabled(enabled, title)
        }
    }
}
